### PR TITLE
Consolidate gpt continue training script into pretraining script

### PR DESCRIPTION
**Pull request by yaoyu-33**
# What does this PR do ?

## Continual Learning with Pretrained Checkpoints

Continual learning enables large language models (LLMs) to acquire new skills and remain current with the swiftly advancing landscape of human knowledge. This guide details how to engage in continual learning using existing pretrained checkpoints with the `examples/nlp/language_modeling/megatron_gpt_pretraining.py` script. This process is applicable to various models including Llama 1/2/3, Gemma, Mistral, Mixtral, and others.

### Configuration for Continual Learning
To enable continual learning, modify the `model` configuration in your script as follows:

```yaml
model:
  # Use the following settings to specify the source for continual learning:
  restore_from_path: null  # Set this to a .nemo file path to restore only the model weights.
  restore_from_ckpt: null  # Set this to a checkpoint path to restore both model weights and optimizer states.
```

- **restore_from_path**: Use this when you want to restore only the model weights from a `.nemo` file.
- **restore_from_ckpt**: Use this to restore both model weights and optimizer states from a PyTorch checkpoint.

### Adjusting Training Configurations
Upon engaging in continual learning, it is possible and often beneficial to modify various training configurations:
- **Model Parallelism**: Depending on the computational resources available, you may adjust the model's parallelism settings.
- **Data Blend**: You can change the dataset or modify how data is blended during training to better suit the new training objectives.
- **Learning Rate Scheduler**: Adjusting the learning rate schedule can help optimize training for the new conditions.

Refer to the *pretrain* section of the documentation for details on configuring these parameters.

### Important Notes on Continual Learning
- **State Reset**: When engaging in continual learning, all states including model parallelism configuration, dataset, learning rate scheduler, and randomness are reset, except for the states loaded from the checkpoint.
- **Resuming Training**: If training is interrupted (e.g., due to a system failure) and you wish to resume the continual learning process, ensure that the restore flags (`restore_from_path` or `restore_from_ckpt`) are set back to `null`. This prevents the system from resetting the states again, which could lead to incorrect results.

**Collection**: [nlp]

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

**Comment by ericharper** [comment]
Can we add github CI test for this and then also test the regular GPT pretraining, with resuming, is the same before and after this PR?

**Comment by ericharper** [comment]
We should also add docs for this, the readme here is a good start.

**Comment by yaoyu-33** [comment]
sure i will add some tests. the gpt resume behavior is not changed in my local tests

File: nemo/collections/nlp/parts/nlp_overrides.py [hunk 1]
```diff
@@ -510,10 +510,25 @@ def load_checkpoint(self, checkpoint_path: Union[str, Path]) -> Dict[str, Any]:
...
             checkpoint['optimizer_states'] = [self.optimizer_sharded_state_dict(is_loading=True)]
-
             if self._check_param_groups_mismatch(checkpoint_path, checkpoint):
-                return self._fix_param_groups(checkpoint_path, checkpoint)
-            return self.checkpoint_io.load_checkpoint(checkpoint_path, sharded_state_dict=checkpoint)
+                checkpoint = self._fix_param_groups(checkpoint_path, checkpoint)
+            else:
+                checkpoint = self.checkpoint_io.load_checkpoint(checkpoint_path, sharded_state_dict=checkpoint)
+
+            if getattr(self.lightning_module, 'continue_training'):
```
Comment (mikolajblaz) [inline nemo/collections/nlp/parts/nlp_overrides.py:518]: Is the intention here to preserve the current (pre-load) optimizer state?
Can you add a comment and maybe extract it to a separate method?

**Comment by yaoyu-33** [comment]
@ericharper  as I work on CI. I did some local tests 
image

File: nemo/collections/nlp/parts/nlp_overrides.py (hunk 1 above)
Comment (yaoyu-33) [reply nemo/collections/nlp/parts/nlp_overrides.py:518]: This is the CT method we leveraged in megatron-lm. Try to support the same in NeMo as well. I will extract it to a separate method.

File: examples/nlp/language_modeling/megatron_gpt_continue_training.py [hunk 2]
```diff
@@ -1,198 +0,0 @@
...
-            app_state.pipeline_model_parallel_split_rank,
-            app_state.virtual_pipeline_model_parallel_rank,
-        ) = fake_initialize_model_parallel(
-            world_size=app_state.model_parallel_size,
-            rank=trainer.global_rank,
-            tensor_model_parallel_size_=cfg.model.tensor_model_parallel_size,
-            pipeline_model_parallel_size_=cfg.model.pipeline_model_parallel_size,
-            pipeline_model_parallel_split_rank_=cfg.model.pipeline_model_parallel_split_rank,
-        )
-    checkpoint_path = inject_model_parallel_rank(
```
Comment (ryxli) [inline examples/nlp/language_modeling/megatron_gpt_continue_training.py:111]: Wondering about the legacy tp_pp partition checkpoint. With this refactor, I don't see a clear way to resume from a checkpoint in this format if we do not save the .nemo file?

```
checkpoint/
  tp_rank_00_pp_rank_000/
    megatron_gpt--....-consumed_samples=1000.ckpt
  tp_rank_00_pp_rank_001/
  ...
  tp_rank_01_pp_rank_000/
  tp_rank_01_pp_rank_001/
  ...
```

Comment (yaoyu-33) [reply examples/nlp/language_modeling/megatron_gpt_continue_training.py:111]: @ryxli it will still work. It uses the lightning resume for ckpt checkpoints (the logic is in NLPoverride), so both legacy and dist format will work. This part of code is redundant anyways.
https://github.com/NVIDIA/NeMo/blob/bbbe826aa524d2ef3d841e50ec97788322a31f4e/nemo/collections/nlp/parts/nlp_overrides.py#L521

**Review by mikolajblaz — APPROVED** [review]

**Review by mikolajblaz — APPROVED** [review]
